File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    # 좌표 2배 확장
    board = [[0] * 102 for _ in range(102)]
    
    # 1. 사각형 경계 그리기
    for x1, y1, x2, y2 in rectangle:
        x1 *= 2; y1 *= 2; x2 *= 2; y2 *= 2
        
        for x in range(x1, x2 + 1):
            for y in range(y1, y2 + 1):
                board[x][y] = 1
    
    # 2. 내부 지우기
    for x1, y1, x2, y2 in rectangle:
        x1 *= 2; y1 *= 2; x2 *= 2; y2 *= 2
        
        for x in range(x1 + 1, x2):
            for y in range(y1 + 1, y2):
                board[x][y] = 2
    
    # BFS 준비
    q = deque()
    q.append((characterX * 2, characterY * 2, 0))
    visited = [[False] * 102 for _ in range(102)]
    
    # 4방향
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
    
    while q:
        x, y, dist = q.popleft()
        
        if x == itemX * 2 and y == itemY * 2:
            return dist // 2
        
        if visited[x][y]:
            continue
        visited[x][y] = True
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx < 102 and 0 <= ny < 102:
                if board[nx][ny] == 1 and not visited[nx][ny]:
                    q.append((nx, ny, dist + 1))

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.java
import java.util.*;

class Solution {
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        // 일단 테두리를 따라가면 안되니까 2배로 만들고 안에 비우기
        int[][] board = new int[102][102];
        
        for(int[] r : rectangle){
            int x1 = r[0] * 2; int y1 = r[1] * 2; int x2 = r[2] * 2; int y2 = r[3] * 2;
            
            for(int i = x1; i < x2 + 1; i++){
                for(int j = y1; j < y2 + 1; j++){
                    board[i][j] = 2;
                }
            }
        }
        for(int[] r : rectangle){
            int x1 = r[0] * 2; int y1 = r[1] * 2; int x2 = r[2] * 2; int y2 = r[3] * 2;
            
            for(int i = x1 + 1; i < x2; i++){
                for(int j = y1 + 1; j < y2; j++){
                    board[i][j] = 1;
                }
            }
        }
        
        
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{characterX * 2, characterY * 2, 0});
        boolean[][] v = new boolean[102][102];
        
        int answer = 0;
        int[][] d = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        
        while(!q.isEmpty()){
            int[] now = q.poll();
            
            if(now[0] == itemX * 2 && now[1] == itemY * 2) return now[2] / 2;
            
            for(int i = 0; i < 4; i ++){
                int nx = now[0] + d[i][0];
                int ny = now[1] + d[i][1];
            
                if(nx < 0 || nx > 101 || ny < 0 || ny > 101) continue;
                if(board[nx][ny] == 2 && !v[nx][ny]){
                    v[nx][ny] = true;
                    q.offer(new int[]{nx, ny, now[2] + 1});
                }
            }
            
        }
        return answer;
    }
}